### PR TITLE
Refactor usage of global `Homebrew.args`.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

---

Continuation of https://github.com/Homebrew/brew/pull/8068.

### CODE REVIEW

File: Library/Homebrew/install.rb [hunk 1]
```diff
@@ -40,11 +38,11 @@ def attempt_directory_creation
     end
 
     def check_cc_argv
-      return unless Homebrew.args.cc
+      return unless (cc = args.cc)
```
Comment (claui) [inline Library/Homebrew/install.rb:41]: Did you mean `==`?

Comment (reitermarkus) [reply Library/Homebrew/install.rb:41]: No, RuboCop actually makes you use parentheses for this specific case.

**Comment by issyl0** [comment]
Looks like `fetch.rb` also needs some work. From Sorbet on the latest `master`:

```
fetch.rb:6: Method args does not exist on Homebrew::Fetch https://srb.help/7003
     6 |      return true if args.force_bottle? && f.bottle
                             ^^^^

fetch.rb:8: Method args does not exist on Homebrew::Fetch https://srb.help/7003
     8 |      return false if args.build_formula_from_source?(f)
                              ^^^^
```

File: Library/Homebrew/install.rb (hunk 1 above)
Comment (claui) [reply Library/Homebrew/install.rb:41]: Using the return value of `=` is clever but makes the code hard to reason about.

I know what the return value is in C but I don’t know it in Ruby. It’s basically impossible to look it up on https://ruby-doc.org.

File: Library/Homebrew/brew.rb [hunk 2]
```diff
@@ -159,8 +157,6 @@ def output_unsupported_error
   onoe e
   $stderr.puts e.backtrace if Homebrew.args.debug?
 
-  output_unsupported_error if Homebrew.args.HEAD?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.rb:162]: Still want this output.

File: Library/Homebrew/cmd/gist-logs.rb [hunk 3]
```diff
@@ -142,8 +144,8 @@ def create_issue(repo, title, body)
   def gist_logs
     gist_logs_args.parse
 
-    Install.perform_preinstall_checks(all_fatal: true)
-    Install.perform_build_from_source_checks(all_fatal: true)
+    perform_preinstall_checks(all_fatal: true)
+    perform_build_from_source_checks(all_fatal: true)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/gist-logs.rb:148]: The previous version reads better to me.

File: Library/Homebrew/cmd/install.rb [hunk 4]
```diff
@@ -255,17 +256,17 @@ def install
 
     return if formulae.empty?
 
-    Install.perform_preinstall_checks
+    perform_preinstall_checks
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/install.rb:259]: I like knowing what module this came from.

File: Library/Homebrew/cmd/reinstall.rb [hunk 5]
```diff
@@ -58,22 +61,22 @@ def reinstall
 
     FormulaInstaller.prevent_build_flags unless DevelopmentTools.installed?
 
-    Install.perform_preinstall_checks
+    perform_preinstall_checks
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/reinstall.rb:64]: I like knowing which module this came from

File: Library/Homebrew/cmd/upgrade.rb [hunk 6]
```diff
@@ -60,7 +62,7 @@ def upgrade
 
     FormulaInstaller.prevent_build_flags unless DevelopmentTools.installed?
 
-    Install.perform_preinstall_checks
+    perform_preinstall_checks
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/upgrade.rb:65]: I like knowing which module this came from

File: Library/Homebrew/formula.rb [hunk 7]
```diff
@@ -2363,6 +2369,15 @@ def build
...
     end
 
+    # @private
+    def build_flags
+      mod_name = to_s.split("::")[0..-2].join("::")
+      return [] if mod_name.empty?
+
+      mod = const_get(mod_name)
+      mod.const_get(:BUILD_FLAGS)
+    end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:2379]: What's going on here and why wasn't it needed before? I find this pretty opaque.

File: Library/Homebrew/formula.rb [hunk 8]
```diff
@@ -273,7 +279,7 @@ def validate_attributes!
   # and is specified to this instance.
   def installed_alias_path
     path = build.source["path"] if build.is_a?(Tab)
-    return unless path&.match?(%r{#{HOMEBREW_TAP_DIR_REGEX}/Aliases})
+    return unless path&.to_s&.match?(%r{#{HOMEBREW_TAP_DIR_REGEX}/Aliases})
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:282]: Can this be split into three `return unless`?

File: Library/Homebrew/formula_installer.rb [hunk 9]
```diff
@@ -95,10 +97,10 @@ def self.clear_installed
   # it's necessary to interrupt the user before any sort of installation
   # can proceed. Only invoked when the user has no developer tools.
   def self.prevent_build_flags
-    build_flags = Homebrew.args.collect_build_args
-    return if build_flags.empty?
+    build_flags = args&.collect_build_args
+    return unless build_flags.present?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb:101]:
```suggestion
    return if build_flags.blank?
```

File: Library/Homebrew/formula_installer.rb [hunk 10]
```diff
@@ -144,7 +146,7 @@ def pour_bottle?(install_bottle_options = { warn: false })
 
   def install_bottle_for?(dep, build)
     return pour_bottle? if dep == formula
-    return false if Homebrew.args.build_formula_from_source?(dep)
+    return false if args&.build_formula_from_source?(dep)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb:149]: I don't like that there's a bunch of `nil` `args` around now. I think a `NullArgs` object or something would make the API nicer.

File: Library/Homebrew/formulary.rb [hunk 11]
```diff
@@ -27,18 +27,37 @@ def self.formula_class_get(path)
...
     raise "Formula loading disabled by HOMEBREW_DISABLE_LOAD_FORMULA!" if Homebrew::EnvConfig.disable_load_formula?
 
     mod = Module.new
     const_set(namespace, mod)
+
+    class_name = class_s(name)
+
     begin
-      mod.module_eval(contents, path)
+      if contents.respond_to?(:call)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb:39]: Why is this needed now and wasn't before?

File: Library/Homebrew/install.rb (hunk 1 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/install.rb:41]:
```suggestion
      cc = args.cc
      return unless cc
```

File: Library/Homebrew/patch.rb [hunk 12]
```diff
@@ -159,6 +159,11 @@ def apply
         end
       end
     end
+  rescue ErrorDuringExecution => e
+    raise unless (f = resource.owner&.owner)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/patch.rb:163]:
```suggestion
    f = resource.owner&.owner
    raise unless f
```

File: Library/Homebrew/test/support/fixtures/testball_bottle.rb [hunk 13]
```diff
@@ -1,14 +1,15 @@
...
+  def initialize(name = "testball_bottle", path = Pathname.new(__FILE__).expand_path, spec = :stable,
+                 alias_path: nil, force_bottle: false)
     self.class.instance_eval do
       stable.url "file://#{TEST_FIXTURE_DIR}/tarballs/testball-0.1.tbz"
       stable.sha256 TESTBALL_SHA256
       stable.bottle do
         cellar :any_skip_relocation
         root_url "file://#{TEST_FIXTURE_DIR}/bottles"
-        sha256 "d48bbbe583dcfbfa608579724fc6f0328b3cd316935c6ea22f134610aaf2952f" => Utils::Bottles.tag
+        sha256 "8f9aecd233463da6a4ea55f5f88fc5841718c013f3e2a7941350d6130f1dc149" => Utils::Bottles.tag
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/support/fixtures/testball_bottle.rb:12]: Why does this change?

File: Library/Homebrew/brew.rb (hunk 2 above)
Comment (reitermarkus) [reply Library/Homebrew/brew.rb:162]: This is handled by making failed patches raise a `BuildError`.

File: Library/Homebrew/cmd/gist-logs.rb (hunk 3 above)
Comment (reitermarkus) [reply Library/Homebrew/cmd/gist-logs.rb:148]: The previous version will always use the global `Homebrew.args`, this will use the `args` in the current class/module. This just happens to still be `Homebrew.args` at the moment.

File: Library/Homebrew/cmd/install.rb (hunk 4 above)
Comment (reitermarkus) [reply Library/Homebrew/cmd/install.rb:259]: See https://github.com/Homebrew/brew/pull/8084#discussion_r460737938.

File: Library/Homebrew/cmd/reinstall.rb (hunk 5 above)
Comment (reitermarkus) [reply Library/Homebrew/cmd/reinstall.rb:64]: See https://github.com/Homebrew/brew/pull/8084#discussion_r460737938.

File: Library/Homebrew/cmd/upgrade.rb (hunk 6 above)
Comment (reitermarkus) [reply Library/Homebrew/cmd/upgrade.rb:65]: See https://github.com/Homebrew/brew/pull/8084#discussion_r460737938.

File: Library/Homebrew/test/support/fixtures/testball_bottle.rb (hunk 13 above)
Comment (reitermarkus) [reply Library/Homebrew/test/support/fixtures/testball_bottle.rb:12]: The `testball_bottle.rb` inside the bottle `.tar.gz` needed to be changed as well.

File: Library/Homebrew/install.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/install.rb:41]: I think this pattern is quite idiomatic Ruby.

Comment (claui) [reply Library/Homebrew/install.rb:41]: > I think this pattern is quite idiomatic Ruby.

You’re right, it’s [used quite widely](https://github.com/rubocop-hq/ruby-style-guide/tree/0fe5678a01618ce87842b4868363419cb89183d9#safe-assignment-in-condition) but there are people who rightfully oppose it as well ([example](https://gitlab.com/gitlab-org/gitlab-foss/-/issues/45605)).

I think it’s clever but needlessly confusing and [difficult to understand](https://github.com/Homebrew/brew/pull/8084#discussion_r460579310).

File: Library/Homebrew/formula.rb (hunk 7 above)
Comment (reitermarkus) [reply Library/Homebrew/formula.rb:2379]: Previously, the flags were all global in `Homebrew.args`. Here we set a constant inside the formula namespace with the passed down flags. That's really the best we can do here. It is still not really as scoped as it should be, and it can't be, since the flags already need to be available when the `Formula` subclass is evaluated. Simply put, you cannot load the same formula with different build flags.

File: Library/Homebrew/cmd/gist-logs.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/gist-logs.rb:148]: Hmm, I think I'd prefer keeping it in the module and passing through `args`?

File: Library/Homebrew/install.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/install.rb:41]: Agreed with @claui but not a blocker.

File: Library/Homebrew/test/support/fixtures/testball_bottle.rb (hunk 13 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/support/fixtures/testball_bottle.rb:12]: Why? Just so I can understand.

Comment (reitermarkus) [reply Library/Homebrew/test/support/fixtures/testball_bottle.rb:12]: It overrides the formula's `initialize` method, which needed another argument added.

File: Library/Homebrew/formulary.rb (hunk 11 above)
Comment (reitermarkus) [reply Library/Homebrew/formulary.rb]: Leftover from debugging. Removed.

File: Library/Homebrew/cmd/gist-logs.rb (hunk 3 above)
Comment (reitermarkus) [reply Library/Homebrew/cmd/gist-logs.rb]: Done.

File: Library/Homebrew/install.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/install.rb]: This check isn't needed anymore now.

File: Library/Homebrew/patch.rb (hunk 12 above)
Comment (reitermarkus) [reply Library/Homebrew/patch.rb]: This check isn't needed anymore now.

File: Library/Homebrew/formula_installer.rb [hunk 14]
```diff
@@ -780,12 +777,6 @@ def sanitized_argv_options
       args << "--devel"
     end
 
-    formula.options.each do |opt|
-      name = opt.name[/^([^=]+)=$/, 1]
-      value = self.args&.value(name) if name
-      args << "--#{name}=#{value}" if value
-    end
-
```
Comment (reitermarkus) [inline Library/Homebrew/formula_installer.rb:771]: Not sure about this one, I guess this was previously used inside formulae using `ARGV.value`, which doesn't work anymore. Am I missing something here?

Comment (MikeMcQuaid) [reply Library/Homebrew/formula_installer.rb:771]: I'm OK with deleting it if you are and we can re-add if it causes issues.

File: Library/Homebrew/formula.rb (hunk 7 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/formula.rb:2388]: Ok, with that in mind can each line be commented more heavily and use some more meaningful variable names? Pretty hard to understand what's going on here without it.

File: Library/Homebrew/formulary.rb [hunk 15]
```diff
@@ -121,22 +123,23 @@ def initialize(name, path)
     #
     # `alias_path` can be overridden here in case an alias was used to refer to
     # a formula that was loaded in another way.
-    def get_formula(spec, alias_path: nil)
-      klass.new(name, path, spec, alias_path: alias_path || self.alias_path)
+    def get_formula(spec, alias_path: nil, force_bottle: false, flags: [])
+      klass(flags: flags)
+        .new(name, path, spec, alias_path: alias_path || self.alias_path, force_bottle: force_bottle)
     end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb:129]:
```suggestion
    def get_formula(spec, alias_path: self.alias_path, force_bottle: false, flags: [])
      klass(flags: flags).new(name, path, spec, alias_path: alias_path, force_bottle: force_bottle)
    end
```

File: Library/Homebrew/formula_installer.rb (hunk 14 above)
Comment (reitermarkus) [reply Library/Homebrew/formula_installer.rb:771]: Yeah, I think we can come up with a better way to refactor this *if* this still used and the way it is used.

File: Library/Homebrew/formulary.rb (hunk 15 above)
Comment (reitermarkus) [reply Library/Homebrew/formulary.rb:129]: Seems like this doesn't quite do what you would expect.

Comment (reitermarkus) [reply Library/Homebrew/formulary.rb:129]: Some other methods explicitly pass `alias_path: nil`, overriding it if it is in the signature.

Comment (MikeMcQuaid) [reply Library/Homebrew/formulary.rb:129]: Doh. 
```suggestion
    def get_formula(spec, alias_path: nil, force_bottle: false, flags: [])
      alias_path ||= self.alias_path
      klass(flags: flags).new(name, path, spec, alias_path: alias_path, force_bottle: force_bottle)
    end
```
